### ISSUE TITLE
Enhancement: [await-thenable] should prohibit using a sync disposable with `await using`

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#10208]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/await-thenable/, https://typescript-eslint.io/rules/no-floating-promises, https://typescript-eslint.io/rules/no-misused-promises

### Description

I would like to find a place to lint for providing a sync disposable to an `await using` statement. This relates closely to https://github.com/typescript-eslint/typescript-eslint/issues/8858, so maybe await-thenable is the right place for this check? But it also has concerns very adjacent to no-floating-promises and no-misused-promises, so I'm not sure.

### Fail

```TypeScript
async function unhandledRejection() {             
    await using _ = {
        async [Symbol.dispose]() {
            throw new Error('dispose error');
        }
    }
}
```

### Pass

```TypeScript
async function properlyHandledRejection() {             
    await using _ = {
        async [Symbol.asyncDispose]() {
            throw new Error('dispose error');
        }
    }
}
```

### Additional Info

The promise rejection in `unhandledRejection` cannot be caught, since the `[Symbol.dispose]()` method is called, but _is not awaited_, by an `await using` statement. Only the `[Symbol.asyncDispose]()` method is awaited. 

There's a lot of subtlety here so I've provided a repo to be able to play with variations on this at https://github.com/kirkwaiblinger/typescript-eslint-repro-await-using

**Comment by kirkwaiblinger** [comment]
Note that the fundamental concern here is that a `Promise`-returning `[Symbol.dispose]()` method is a clear vector for unhandled rejections, whether invoked by an `await using` or an ordinary `using` statement.

**Comment by bradzacher** [comment]
probably makes most sense for this to be in `await-thenable`, I think.

**Comment by kirkwaiblinger** [comment]
@bradzacher 
So my thing is, if we do that, how do we also ban 
```ts
function unhandledRejection() {             
    using _ = {
        async [Symbol.dispose]() {
            throw new Error('dispose error');
        }
    }
}
```
It doesn't go in await-thenable, since that's all about unnecessary `await`ing of non-thenables. Maybe no-misused-promises then? Note that there's no `() => Promise<void>` to `() => void` assignment happening here, since the object literal is provided literally to the `using` declaration, therefore, we'd need to add a new check for `using` declarations specifically, rather than being able to rely on the existing checks of no-misused-promises.

**Comment by bradzacher** [comment]
Does the disposable method have a default type definition? If it does then iirc there's another rule that should already catch it.

If it doesn't then no-misused-promises would probably be a good fit - iirc it already handles array iteration callbacks.

**Comment by kirkwaiblinger** [comment]
> Does the disposable method have a default type definition? If it does then iirc there's another rule that should already catch it.
> 
> If it doesn't then no-misused-promises would probably be a good fit - iirc it already handles array iteration callbacks.

After a bit of investigation...

### `await using` sync disposable

```ts
await using x = {
    [Symbol.dispose]() { }
}
```
can be easily added to await-thenable, #10209 in progress.

### Promise returning function assigned to `Disposable` variable.

```ts
const x: Disposable = {
    async [Symbol.dispose]() {}
} 
```
already flagged by no-misused-promises. This is a garden variety `() => Promise<void>` to `() => void` assignment.

### Promise returning function in variable assigned to `Disposable` variable

```ts
const x = {
    async [Symbol.dispose]() {}
} 
const y: Disposable = x;
```
not flagged by current tooling at all. Could/should be added to no-misused-promises (or upcoming strict-void-return, #9707). Arguably this is just an existing hole in the rule for `checksVoidReturn` that it doesn't inspect the properties within a variable being assigned.

### `using x` with variable

```ts
const x = {
    async [Symbol.dispose]() {}
} 
using y = x;
```
not flagged by current tooling at all (unless an explicit `: Disposable` type annotation is provided). Could/should be added to no-misused-promises (or upcoming strict-void-return, #9707).

### `using x` with literal

```ts 
using y = {
    async [Symbol.dispose]() {}
};
```
not flagged by current tooling at all. Could/should be added to no-misused-promises (or upcoming strict-void-return, #9707).

### async disposable class

```ts
class d implements Disposable {
    async [Symbol.dispose]() {}
}
```

Not flagged by current tooling at all. But it _should be_, by `no-misused-promises`'s `inheritedMethods` checks. This is a bug due to the key being a (static) "computed" symbol rather than a string literal.

**Comment by kirkwaiblinger** [comment]
Closing this issue since #10209 is resolved. 

Remaining work is tracked in #10211 and #10357
